### ISSUE TITLE
[CI] Tests of _cat output containing dates can fail as ZonedDateTime.toString() does not have deterministic format

### ISSUE THREAD

**Issue by droberts195** [elastic/elasticsearch#32466]
`cat.indices/10_basic/Test cat indices output` failed with the following error:

```
FAILURE 0.12s | MixedClusterClientYamlTestSuiteIT.test {p0=cat.indices/10_basic/Test cat indices output} <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: Failure at [cat.indices/10_basic:40]: field [$body] was expected to match the provided regex but didn't
   > Expected: ^(
   >    index1                                                    \s+
   >    (\d+)                                                     \s+
   >    (\d\d\d\d\-\d\d\-\d\dT\d\d:\d\d:\d\d.\d\d\dZ) \s+
   >    (\d+)                                                     \s+
   >    (\d\d\d\d\-\d\d\-\d\dT\d\d:\d\d:\d\d.\d\d\dZ) \s*
   >  )
   >  $
   >      but: was "index1 1532913914000 2018-07-30T01:25:14Z 1532913914000 2018-07-30T01:25:14Z\n"
   > 	at __randomizedtesting.SeedInfo.seed([5CE2FDA0291D51B9:D4B6C27A87E13C41]:0)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:394)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.test(ESClientYamlSuiteTestCase.java:371)
   > 	at java.lang.Thread.run(Thread.java:748)
   > Caused by: java.lang.AssertionError: field [$body] was expected to match the provided regex but didn't
   > Expected: ^(
   >    index1                                                    \s+
   >    (\d+)                                                     \s+
   >    (\d\d\d\d\-\d\d\-\d\dT\d\d:\d\d:\d\d.\d\d\dZ) \s+
   >    (\d+)                                                     \s+
   >    (\d\d\d\d\-\d\d\-\d\dT\d\d:\d\d:\d\d.\d\d\dZ) \s*
   >  )
   >  $
   >      but: was "index1 1532913914000 2018-07-30T01:25:14Z 1532913914000 2018-07-30T01:25:14Z\n"
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.test.rest.yaml.section.MatchAssertion.doAssert(MatchAssertion.java:67)
   > 	at org.elasticsearch.test.rest.yaml.section.Assertion.execute(Assertion.java:76)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:387)
   > 	... 37 more
```

The full log is https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.4+bwc-tests/6/console

Then problem is that the regex expects milliseconds but the output only contains seconds.

This does not reproduce easily.  However, by making the following edit the problem will reproduce 50% of the time (presumably because it's a mixed cluster and the code edit only applies to one version):

```
--- a/server/src/main/java/org/elasticsearch/rest/action/cat/RestIndicesAction.java
+++ b/server/src/main/java/org/elasticsearch/rest/action/cat/RestIndicesAction.java
@@ -380,7 +380,7 @@ public class RestIndicesAction extends AbstractCatAction {
             table.addCell(primaryStats.getDocs() == null ? null : primaryStats.getDocs().getDeleted());
 
             table.addCell(indexMetaData.getCreationDate());
-            table.addCell(ZonedDateTime.ofInstant(Instant.ofEpochMilli(indexMetaData.getCreationDate()), ZoneOffset.UTC));
+            table.addCell(ZonedDateTime.ofInstant(Instant.ofEpochSecond(indexMetaData.getCreationDate() / 1000L), ZoneOffset.UTC));
 
             table.addCell(totalStats.getStore() == null ? null : totalStats.getStore().size());
             table.addCell(primaryStats.getStore() == null ? null : primaryStats.getStore().size());
```

So the problem is that `ZonedDateTime.toString()` does not output milliseconds if the time it's storing is an exact number of seconds.

I suspect this problem is new to 6.4 due to the Joda to Java time migration.

**Comment by droberts195** [comment]
Another one, this time on the 6.x branch: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+default-distro-bwc-tests/201/console

**Comment by danielmitterdorfer** [comment]
Another one on 6.x: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+default-distro-bwc-tests/203/console

Reproduction line:

```
./gradlew :qa:mixed-cluster:v6.3.1#mixedClusterTestRunner -Dtests.seed=C3EDE8615C4F4515 -Dtests.class=org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT -Dtests.method="test {p0=cat.indices/10_basic/Test cat indices output}" -Dtests.security.manager=true -Dtests.locale=no-NO -Dtests.timezone=Canada/Pacific -Dtests.distribution=zip
```

**Comment by davidkyle** [comment]
Another instance: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+matrix-java-periodic/ES_BUILD_JAVA=java10,ES_RUNTIME_JAVA=java8,nodes=virtual&&linux/220/console

```
java.lang.AssertionError: Failure at [cat.indices/10_basic:40]: field [$body] was expected to match the provided regex but didn't
Expected: ^(
   index1                                                    \s+
   (\d+)                                                     \s+
   (\d\d\d\d\-\d\d\-\d\dT\d\d:\d\d:\d\d.\d\d\dZ) \s+
   (\d+)                                                     \s+
   (\d\d\d\d\-\d\d\-\d\dT\d\d:\d\d:\d\d.\d\d\dZ) \s*
 )
 $
     but: was "index1 1533556765000 2018-08-06T11:59:25Z 1533556765000 2018-08-06T11:59:25Z\n"
```

**Comment by andyb-elastic** [comment]
The fix above causes a problem on 6.4 - I could reproduce it on 6.4 but not master or 6.x. The formatted date it checks ends up with three `Z` time zone designators at the end - I'm not sure if it's intentional but I've not heard of this in iso 8601 before

```
  1> [2018-08-14T15:00:58,144][INFO ][o.e.t.r.IntegTestZipClientYamlTestSuiteIT] [test {p0=cat.indices/10_basic/Test cat indices output}]: after test
FAILURE 0.08s | IntegTestZipClientYamlTestSuiteIT.test {p0=cat.indices/10_basic/Test cat indices output} <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: Failure at [cat.indices/10_basic:40]: field [$body] was expected to match the provided regex but didn't
   > Expected: ^(
   >    index1                                                    \s+
   >    (\d+)                                                     \s+
   >    (\d\d\d\d\-\d\d\-\d\dT\d\d:\d\d:\d\d.\d\d\dZ) \s+
   >    (\d+)                                                     \s+
   >    (\d\d\d\d\-\d\d\-\d\dT\d\d:\d\d:\d\d.\d\d\dZ) \s*
   >  )
   >  $
   >      but: was "index1 1534273258075 2018-08-14T19:00:58.075ZZZ 1534273258075 2018-08-14T19:00:58.075ZZZ\n"
   >    at __randomizedtesting.SeedInfo.seed([96CA3F9159C09648:1E9E004BF73CFBB0]:0)
   >    at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:382)
   >    at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.test(ESClientYamlSuiteTestCase.java:359)
   >    at java.lang.Thread.run(Thread.java:748)
   > Caused by: java.lang.AssertionError: field [$body] was expected to match the provided regex but didn't
   > Expected: ^(
   >    index1                                                    \s+
   >    (\d+)                                                     \s+
   >    (\d\d\d\d\-\d\d\-\d\dT\d\d:\d\d:\d\d.\d\d\dZ) \s+
   >    (\d+)                                                     \s+
   >    (\d\d\d\d\-\d\d\-\d\dT\d\d:\d\d:\d\d.\d\d\dZ) \s*
   >  )
   >  $
   >      but: was "index1 1534273258075 2018-08-14T19:00:58.075ZZZ 1534273258075 2018-08-14T19:00:58.075ZZZ\n"
   >    at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   >    at org.elasticsearch.test.rest.yaml.section.MatchAssertion.doAssert(MatchAssertion.java:67)
   >    at org.elasticsearch.test.rest.yaml.section.Assertion.execute(Assertion.java:76)
   >    at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:375)
   >    ... 37 more
```

```
REPRODUCE WITH: ./gradlew :distribution:archives:integ-test-zip:integTestRunner -Dtests.seed=96CA3F9159C09648 -Dtests.class=org.elasticsearch.test.rest.IntegTestZipClientYamlTestSuiteIT -Dtests.method="test {p0=cat.indices/10_basic/Test cat indices output}" -Dtests.security.manager=true -Dtests.locale=fr-BE -Dtests.timezone=America/Lower_Princes
```

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.4+intake/57/console

[build-57.txt](https://github.com/elastic/elasticsearch/files/2288379/build-57.txt)

**Comment by rjernst** [comment]
The problem was https://github.com/elastic/elasticsearch/pull/32592 was not backported all the way to 6.4. I've done that, and the test passes now. I've re-enabled this test in 3991ca0.
